src/Database/PostgreSQL/PQTypes/Model/Trigger.hs

```diff
@@ -0,0 +1,284 @@
+             && triggerDeferrable t1 == triggerDeferrable t2
+             && triggerInitiallyDeferred t1 == triggerInitiallyDeferred t2
+             && triggerFunction t1 == triggerFunction t2
+
+-- | Make a trigger name that can be used in SQL.
+--
+-- Given a base @name@ and @tableName@, return a new name that will be used as the
+-- actually name of the trigger in an SQL query. The returned name is in the format
+-- @trg\__\<tableName\>\__\<name\>@.
+--
+-- @since 1.15.0
+triggerMakeName :: RawSQL () -> RawSQL () -> RawSQL ()
+triggerMakeName name tableName = "trg__" <> tableName <> "__" <> name
+
+-- | Return the trigger's base name.
+--
+-- Given the trigger's actual @name@ and @tableName@, return the base name of the
+-- trigger. This is basically the reverse of what 'triggerMakeName' does.
+--
+-- @since 1.15.0
+triggerBaseName :: RawSQL () -> RawSQL () -> RawSQL ()
+triggerBaseName name tableName =
+  rawSQL (snd . Text.breakOnEnd (unRawSQL tableName <> "__") $ unRawSQL name) ()
+
+triggerEventName :: TriggerEvent -> RawSQL ()
+triggerEventName = \case
+  TriggerInsert -> "INSERT"
+  TriggerUpdate -> "UPDATE"
+  TriggerDelete -> "DELETE"
+
+-- | Build an SQL statement that creates a trigger.
+--
+-- Only supports @CONSTRAINT@ triggers which can only run @AFTER@.
+--
+-- @since 1.15.0
+sqlCreateTrigger :: Trigger -> RawSQL ()
+sqlCreateTrigger Trigger{..} =
+  "CREATE CONSTRAINT TRIGGER" <+> trgName
+    <+> "AFTER" <+> trgEvents
+    <+> "ON" <+> triggerTable
+    <+> trgTiming
+    <+> "FOR EACH ROW"
+    <+> trgWhen
+    <+> "EXECUTE FUNCTION" <+> trgFunction
+    <+> "();"
+  where
+    trgName
+      | triggerName == "" = error "Trigger must have a name."
+      | otherwise = triggerMakeName triggerName triggerTable
+    trgEvents
+      | triggerEvents == Set.empty = error "Trigger must have at least one event."
+      | otherwise = mintercalate " OR " . map triggerEventName $ Set.toList triggerEvents
+    trgTiming = let deferrable = (if triggerDeferrable then "" else "NOT") <+> "DEFERRABLE"
+                    deferred   = if triggerInitiallyDeferred
+                                 then "INITIALLY DEFERRED"
+                                 else "INITIALLY IMMEDIATE"
+                in deferrable <+> deferred
+    trgWhen = maybe "" ("WHEN" <+>) triggerWhen
+    trgFunction = tfName triggerFunction
+
+-- | Get all noninternal triggers from the database.
+--
+-- Run a query that returns all database triggers marked as @tgisinternal = false@.
+--
+-- Note that, in the background, to get the trigger's @WHEN@ clause and the source code of
+-- the attached function, the entire query that had created the trigger is received using
+-- @pg_get_triggerdef(t.oid)::text@ and then parsed. The result of that call will be
+-- decompiled and normalized, which means that it's likely not what the user had
+-- originally actually typed. Therefore, 'triggerWhen' and 'tfSource' of 'triggerFunction'
```

Comment:
This comment is for getting the triggers out of the database and warns the potential users. I don't think we actually have to compare the `WHEN` clauses, see the comment in `instance Eq Trigger`.

Reply:
We talked about it, let's try to compare these clauses for higher consistency guarantees.

Reply:
Having done a few experiments, I don't think it's safe to add the comparison and, actually, I think that the code that does the same with CHECKs is essentially broken.

Consider this:
```
Trigger { ...
        , triggerWhen = Just " regexp_replace(NEW.first_name, 'H[a-z]llo(.*):)', 'World,\\1') !=OlD.last_name  "
          ...
        }
```
This is a valid WHEN condition for an `AFTER UPDATE` trigger. Including the weird cases in `NEW` and `OlD`. When you get that condition from the database (using `pg_get_triggerdef()`; for checks, the code uses `pg_get_constraintdef()` which I guess does pretty much the same thing, only for constraints), you get the following:
```
db=# select pg_get_triggerdef(129371,true);
... WHEN (regexp_replace(new.first_name, 'H[a-z]llo(.*):)'::text, 'World,\1'::text) <> old.last_name) ...
```
Notice several things:
* There are casts added (`::text`) in some places. How would you determine where to add those and of which type during the comparison?
* The case of the implicit variables changes. You can't lowercase the entire WHEN, or only parts that would match 'old' and 'new' in a case-insensitive manner, because you might change any string literals in the rest of the condition.
* `!=` changes to `<>`. Both are valid. There might be more "aliases" like this. While you could make a translation table of sorts and substitute them, you would be depending on the postgres internals, which is a bad idea.

The point is that the comparison is impossible to do in a way that would make it usable in more cases than just very trivial ones. And for that reason, I think it's actually more harmful than useful.

The example is silly, but it gets the point across.

Reply:
Yes, I know PostgreSQL normalizes things. But the only thing you need to do to solve any such difference is to copy-paste what it outputs and put this into the code (just as the hint in `checkChecks` suggests). Why is this a problem?

Reply:
The problem seems to be that my assumption was wrong and I have misunderstood the thing. I think I get it know.

So the workflow goes something like this:
1) I create a trigger migration with a WHEN clause in it.
2) I run the migration framework with this migration and the code fails saying that then WHEN clauses are different. It gives a hint just like `checkChecks` does.
3) I "review" the migration and rewrite the WHEN clause in the code so that it matches the normalized version given by the db.
4) I now have a valid consistency check. If someone changes the trigger manually in the DB, at least I know about it.

Is this correct?

Reply:
just a quick note: AFAIK, there is no warranty that this will return the exact same string across versions of Postgres

Reply:
> just a quick note: AFAIK, there is no warranty that this will return the exact same string across versions of Postgres

That's what I was poorly trying to communicate with "you would be depending on the postgres internals, which is a bad idea." It's a potential source of future issues. On the other hand, I understand that without it, there isn't a way to check if someone changed the trigger's WHEN condition manually. So I guess the additional consistency check overrules the potential future issues.

Reply:
> there is no warranty that this will return the exact same string across versions of Postgres

That's fine, we can deal with that if that issue arises.

Reply:
@jsynacek just to confirm, your 4 points are correct 👍